File: plugin/zeppelin/src/main/scala/org/intellij/plugin/zeppelin/service/execution/ZeppelinExecutionManager.scala
package org.intellij.plugin.zeppelin.service.execution

import java.util.concurrent.ConcurrentLinkedQueue

import org.intellij.plugin.zeppelin.api.ZeppelinApi
import org.intellij.plugin.zeppelin.api.websocket._
import org.intellij.plugin.zeppelin.constants.ZeppelinConstants
import org.intellij.plugin.zeppelin.models.ExecuteContext
import org.intellij.plugin.zeppelin.utils.ZeppelinLogger

/**
  * Service which execute code on Zeppelin and manager queue of execution tasks and messages
  *
  * @param api - an instance of zeppelin API
  */
class ZeppelinExecutionManager(api: ZeppelinApi, executionHandlerFactory: ExecutionHandlerFactory) {
  val tasks = new ConcurrentLinkedQueue[TaskExecutor]()
  init()

  def execute(executeContext: ExecuteContext): Unit = {
    ZeppelinLogger.printMessage(ZeppelinConstants.PARAGRAPH_RUNNED.format(executeContext.text))
    val task = TaskExecutor(executeContext.paragraphId, executionHandlerFactory.create())
    tasks.add(task)
    api.runCode(executeContext)
    clean()
  }

  private def clean(): Unit = {
    var task = tasks.peek()
    while (task != null && task.isCompleted.get()) {
      tasks.poll()
      task = tasks.peek()
    }
  }

  private def getOrCreateTaskExecutor(id: String): Option[TaskExecutor] = {
    val tasksArray = tasks.toArray[TaskExecutor](new Array[TaskExecutor](0))
    tasksArray.find(_.id == id)
  }

  private def handleParagraph(message: ResponseMessage): Unit = {
    val paragraphResponse = ExecutionModelConverter.getParagraphResponse(message.data)
    val taskExecutor = getOrCreateTaskExecutor(paragraphResponse.id)
    if (taskExecutor.isEmpty) return
    taskExecutor.get.paragraph(paragraphResponse)
  }

  private def handleParagraphAppendOutput(message: ResponseMessage): Unit = {
    val updateOutput = ExecutionModelConverter.getOutputResult(message.data)
    val taskExecutor = getOrCreateTaskExecutor(updateOutput.paragraphId)
    if (taskExecutor.isEmpty) return
    taskExecutor.get.appendOutput(updateOutput)
  }

  private def handleParagraphProgress(message: ResponseMessage): Unit = {
    val progress = ExecutionModelConverter.getProgressResponse(message.data)
    val taskExecutor = getOrCreateTaskExecutor(progress.id)
    if (taskExecutor.isEmpty) return
    taskExecutor.get.progress(progress)
  }

  private def handleParagraphUpdateOutput(message: ResponseMessage): Unit = {
    val updateOutput = ExecutionModelConverter.getOutputResult(message.data)
    val taskExecutor = getOrCreateTaskExecutor(updateOutput.paragraphId)
    if (taskExecutor.isEmpty) return
    taskExecutor.get.updateOutput(updateOutput)
  }

  private def init(): Unit = {
    api.registerHandler(ResponseCode.PROGRESS, (result: ResponseMessage) => handleParagraphProgress(result))
    api.registerHandler(ResponseCode.PARAGRAPH, (result: ResponseMessage) => handleParagraph(result))
    api.registerHandler(ResponseCode.PARAGRAPH_APPEND_OUTPUT,
      (result: ResponseMessage) => handleParagraphAppendOutput(result))
    api.registerHandler(ResponseCode.PARAGRAPH_UPDATE_OUTPUT,
      (result: ResponseMessage) => handleParagraphUpdateOutput(result))

  }
}

object ZeppelinExecutionManager {
  def apply(api: ZeppelinApi, executionHandlerFactory: ExecutionHandlerFactory): ZeppelinExecutionManager = {
    new ZeppelinExecutionManager(api, executionHandlerFactory)
  }
}

File: zeppelin/src/org/intellij/plugin/zeppelin/service/execution/ZeppelinExecutionManager.kt
package org.intellij.plugin.zeppelin.service.execution

import org.intellij.plugin.zeppelin.api.ZeppelinApi
import org.intellij.plugin.zeppelin.api.websocket.MessageHandler
import org.intellij.plugin.zeppelin.api.websocket.ResponseCode
import org.intellij.plugin.zeppelin.api.websocket.ResponseMessage
import org.intellij.plugin.zeppelin.constants.ZeppelinConstants
import org.intellij.plugin.zeppelin.models.ExecuteContext
import org.intellij.plugin.zeppelin.utils.ZeppelinLogger
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * Service which execute code on Zeppelin and manager queue of execution tasks and messages
 *
 * @param api - an instance of zeppelin API
 */
open class ZeppelinExecutionManager(private val api: ZeppelinApi,
                                    private val executionHandlerFactory: ExecutionHandlerFactory) {
    init {
        api.registerHandler(ResponseCode.PROGRESS, object : MessageHandler {
            override fun handle(result: ResponseMessage) {
                handleParagraphProgress(result)
            }
        })

        api.registerHandler(ResponseCode.PARAGRAPH, object : MessageHandler {
            override fun handle(result: ResponseMessage) {
                handleParagraph(result)
            }
        })

        api.registerHandler(ResponseCode.PARAGRAPH_APPEND_OUTPUT, object : MessageHandler {
            override fun handle(result: ResponseMessage) {
                handleParagraphAppendOutput(result)
            }
        })

        api.registerHandler(ResponseCode.PARAGRAPH_UPDATE_OUTPUT, object : MessageHandler {
            override fun handle(result: ResponseMessage) {
                handleParagraphUpdateOutput(result)
            }
        })
    }

    private val tasks: ConcurrentLinkedQueue<TaskExecutor> = ConcurrentLinkedQueue()

    fun execute(executeContext: ExecuteContext) {
        ZeppelinLogger.printMessage(ZeppelinConstants.PARAGRAPH_RUNNED.format(executeContext.text))
        val task = TaskExecutor(executeContext.paragraphId, executionHandlerFactory.create())
        tasks.add(task)
        api.runCode(executeContext)
        clean()
    }

    private fun clean() {
        var task: TaskExecutor? = tasks.peek()
        while (task != null && task.isCompleted().get()) {
            tasks.poll()
            task = tasks.peek()
        }
    }

    private fun getOrCreateTaskExecutor(id: String): TaskExecutor? {
        val tasksArray: Array<TaskExecutor> = tasks.toArray<TaskExecutor>(arrayOf())
        return tasksArray.find { it.id == id }
    }

    private fun handleParagraph(message: ResponseMessage) {
        val paragraphResponse = ExecutionModelConverter.getParagraphResponse(message.data)
        val taskExecutor: TaskExecutor = getOrCreateTaskExecutor(paragraphResponse.id) ?: return
        taskExecutor.paragraph(paragraphResponse)
    }

    private fun handleParagraphAppendOutput(message: ResponseMessage) {
        val updateOutput = ExecutionModelConverter.getOutputResult(message.data)
        val taskExecutor: TaskExecutor = getOrCreateTaskExecutor(updateOutput.paragraphId) ?: return
        taskExecutor.appendOutput(updateOutput)
    }

    private fun handleParagraphProgress(message: ResponseMessage) {
        val progress = ExecutionModelConverter.getProgressResponse(message.data)
        val taskExecutor: TaskExecutor = getOrCreateTaskExecutor(progress.id) ?: return
        taskExecutor.progress(progress)
    }

    private fun handleParagraphUpdateOutput(message: ResponseMessage) {
        val updateOutput = ExecutionModelConverter.getOutputResult(message.data)
        val taskExecutor: TaskExecutor = getOrCreateTaskExecutor(updateOutput.paragraphId) ?: return
        taskExecutor.updateOutput(updateOutput)
    }
}